File: Week2/week2.js
// Task 1
function findAndPrint(messages, currentStation) {
  let stations = [
    "Songshan",
    "Nanjing Sanmin",
    "Taipei Arena",
    "Nanjing Fuxing",
    "Songjiang Nanjing",
    "Zhongshan",
    "Beimen",
    "Ximen",
    "Xiaonanmen",
    "Chiang Kai-Shek Memorial Hall",
    "Guting",
    "Taipower Building",
    "Gongguan",
    "Wanlong",
    "Jingmei",
    "Dapinglin",
    "Qizhang",
    "Xiaobitan",
    "Xindian City Hall",
    "Xindian",
  ];

  // 取出messages中每人所在捷運站，作為新物件的值
  let newMessages = {};
  let keys = Object.keys(messages);

  for (let i = 0; i < keys.length; i++) {
    let key = keys[i];
    let value = messages[key];
    for (let j = 0; j < stations.length; j++) {
      if (value.includes(stations[j])) {
        value = stations[j];
        newMessages[key] = value;
      }
    }
  }

  // 找最短距離的朋友
  let minDistance = Infinity;
  let closestFriend = null;
  let newKeys = Object.keys(newMessages);

  for (let i = 0; i < newKeys.length; i++) {
    // 切陣列
    let newKey = newKeys[i];
    let newValue = newMessages[newKey];
    let friendLocation = stations.indexOf(newValue);
    let myLocation = stations.indexOf(currentStation);
    let distance = Math.abs(myLocation - friendLocation);

    // 處理支線小碧潭
    if (currentStation == "Xiaobitan" && newValue == "Xindian City Hall") {
      distance++;
    } else if (currentStation == "Xiaobitan" && newValue == "Xindian") {
      distance = distance + 2;
    }

    if (newValue == "Xiaobitan" && currentStation == "Xindian City Hall") {
      distance++;
    } else if (newValue == "Xiaobitan" && currentStation == "Xindian") {
      distance = distance + 2;
    }

    if (distance < minDistance) {
      minDistance = distance;
      closestFriend = newKey;
    }
  }

  console.log(closestFriend);
  return closestFriend;
}

const messages = {
  "Bob": "I'm at Ximen MRT station.",
  "Mary": "I have a drink near Jingmei MRT station.",
  "Copper": "I just saw a concert at Taipei Arena.",
  "Leslie": "I'm at home near Xiaobitan station.",
  "Vivian": "I'm at Xindian station waiting for you.",
};

console.log("--- Task 1 ---");
findAndPrint(messages, "Wanlong"); // print Mary
findAndPrint(messages, "Songshan"); // print Copper
findAndPrint(messages, "Qizhang"); // print Leslie
findAndPrint(messages, "Ximen"); // print Bob
findAndPrint(messages, "Xindian City Hall"); // print Vivian

// Task 2
// 創建全部顧問時間表
function consultantSchedule(consultants) {
  let schedule = {};

  for (let i = 0; i < consultants.length; i++) {
    let name = consultants[i].name;
    schedule[name] = Array(24).fill(0);
  }
  return schedule;
}

function book(consultants, hour, duration, criteria) {
  // 依顧客標準決定顧問排序
  let criteriaResult;

  if (criteria === "price") {
    // 陣列以價錢為標準採升冪排序
    criteriaResult = consultants.sort((a, b) => a.price - b.price);
  } else if (criteria === "rate") {
    // 陣列以評價為標準採降冪排序
    criteriaResult = consultants.sort((a, b) => b.rate - a.rate);
  }

  for (let i = 0; i < consultants.length; i++) {
    let consultantName = criteriaResult[i].name;
    let indieSchedule = schedule[consultantName];
    let count = 0;
    for (let j = hour - 1; j < hour + duration - 1; j++) {
      if (indieSchedule[j] == 1) {
        break;
      } else {
        indieSchedule[j] = 1;
        count++;
      }

      if (count == duration) {
        console.log(consultantName);
        return consultantName;
      }
    }
  }
  console.log("No Service");
  return "No Service";
}

const consultants = [
  { "name": "John", "rate": 4.5, "price": 1000 },
  { "name": "Bob", "rate": 3, "price": 1200 },
  { "name": "Jenny", "rate": 3.8, "price": 800 },
];

let schedule = consultantSchedule(consultants);

console.log("--- Task 2 ---");
book(consultants, 15, 1, "price"); // Jenny
book(consultants, 11, 2, "price"); // Jenny
book(consultants, 10, 2, "price"); // John
book(consultants, 20, 2, "rate"); // John
book(consultants, 11, 1, "rate"); // Bob
book(consultants, 11, 2, "rate"); // No Service
book(consultants, 14, 3, "price"); // John

// Task 3
function func(...data) {
  let middleNameObj = {};

  for (let fullName of data) {
    let index = Math.floor(fullName.length / 2);
    let middleName = fullName[index];
    if (!middleNameObj.hasOwnProperty(middleName)) {
      // 上行用Object.hasOwn(middleNameObj,middleName)此靜態方法是更安全的作法，但因為考量到瀏覽器支援度且在此題非必要故先不用
      middleNameObj[middleName] = [fullName];
    } else {
      middleNameObj[middleName].push(fullName);
    }
  }

  for (let value of Object.values(middleNameObj)) {
    if (value.length === 1) {
      console.log(value.join());
      return value.join();
    }
  }

  console.log("沒有");
  return "沒有";
}

console.log("--- Task 3 ---");
func("彭大牆", "陳王明雅", "吳明"); // print彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花"); // print林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花"); // print沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆"); // print夏曼藍波安

// Task 4
function getNumber(index) {
  if (index === 0) {
    console.log(0);
    return 0;
  }

  let sequence = [0];
  while (sequence.length <= index) {
    let lastElement = sequence[sequence.length - 1];
    if (sequence.length % 3 === 0) {
      sequence.push(lastElement - 1);
    } else {
      sequence.push(lastElement + 4);
    }
  }
  console.log(sequence[sequence.length - 1]);
  return sequence[sequence.length - 1];
}

console.log("--- Task 4 ---");
getNumber(1); // print 4
getNumber(5); // print 15
getNumber(10); // print 25
getNumber(30); // print 70

File: Week2/week2.py
# Task 1
def find_and_print(messages, current_station):
    stations = [
        "Songshan",
        "Nanjing Sanmin",
        "Taipei Arena",
        "Nanjing Fuxing",
        "Songjiang Nanjing",
        "Zhongshan",
        "Beimen",
        "Ximen",
        "Xiaonanmen",
        "Chiang Kai-Shek Memorial Hall",
        "Guting",
        "Taipower Building",
        "Gongguan",
        "Wanlong",
        "Jingmei",
        "Dapinglin",
        "Qizhang",
        "Xiaobitan",
        "Xindian City Hall",
        "Xindian",
    ]

    # 取出messages中每人所在捷運站，作為新物件的值
    new_messages = {}
    for key in messages.keys():
        value = messages[key]
        for j in range(0, len(stations)):
            if stations[j] in value:
                value = stations[j]
                new_messages[key] = value

    # 找最短距離的朋友
    min_distance = float('inf')
    closest_friend = None

    for new_key in new_messages:
        new_value = new_messages[new_key]
        friend_location = stations.index(new_value)
        my_location = stations.index(current_station)
        distance = abs(my_location - friend_location)

        # 處理支線小碧潭
        if current_station == "Xiaobitan" and new_value == "Xindian City Hall":
            distance += 1
        elif current_station == "Xiaobitan" and new_value == "Xindian":
            distance += 2

        if new_value == "Xiaobitan" and current_station == "Xindian City Hall":
            distance += 1
        elif new_value == "Xiaobitan" and current_station == "Xindian":
            distance += 2

        if distance < min_distance:
            min_distance = distance
            closest_friend = new_key

    print(closest_friend)
    return closest_friend


messages = {
    "Leslie": "I'm at home near Xiaobitan station.",
    "Bob": "I'm at Ximen MRT station.",
    "Mary": "I have a drink near Jingmei MRT station.",
    "Copper": "I just saw a concert at Taipei Arena.",
    "Vivian": "I'm at Xindian station waiting for you."
}

print("===Task1===")
find_and_print(messages, "Wanlong")  # print Mary
find_and_print(messages, "Songshan")  # print Copper
find_and_print(messages, "Qizhang")  # print Leslie
find_and_print(messages, "Ximen")  # print Bob
find_and_print(messages, "Xindian City Hall")  # print Vivian


# Task 2
# 創建全部顧問時間表
def consultant_schedule(consultants):
    schedule = {}

    for consultant in consultants:
        name = consultant["name"]
        schedule[name] = [0] * 24
    return schedule


def book(consultants, hour, duration, criteria):
    criteria_result = None

    if criteria == "price":
        criteria_result = sorted(consultants, key=lambda x: x["price"])
    elif criteria == "rate":
        criteria_result = sorted(consultants,
                                 key=lambda x: x["rate"],
                                 reverse=True)

    for consultant in criteria_result:
        consultant_name = consultant["name"]
        indie_schedule = schedule[consultant_name]
        count = 0
        for j in range(hour - 1, hour + duration - 1):
            if indie_schedule[j] == 1:
                break
            else:
                indie_schedule[j] = 1
                count += 1

            if count == duration:
                print(consultant_name)
                return consultant_name

    print("No Service")
    return "No Service"


consultants = [{
    "name": "John",
    "rate": 4.5,
    "price": 1000
}, {
    "name": "Bob",
    "rate": 3,
    "price": 1200
}, {
    "name": "Jenny",
    "rate": 3.8,
    "price": 800
}]

schedule = consultant_schedule(consultants)

print("===Task2===")
book(consultants, 15, 1, "price")  # Jenny
book(consultants, 11, 2, "price")  # Jenny
book(consultants, 10, 2, "price")  # John
book(consultants, 20, 2, "rate")  # John
book(consultants, 11, 1, "rate")  # Bob
book(consultants, 11, 2, "rate")  # No Service
book(consultants, 14, 3, "price")  # John


# Task 3
def func(*data):

    middle_name_dict = {}

    for full_name in data:
        index = len(full_name) // 2
        middle_name = full_name[index]
        if middle_name not in middle_name_dict:
            middle_name_dict[middle_name] = [full_name]
        else:
            middle_name_dict[middle_name].append(full_name)

    for value in middle_name_dict.values():
        if len(value) == 1:
            print(value[0])
            return (value[0])

    print("沒有")
    return "沒有"


print("===Task3===")
func("彭大牆", "陳王明雅", "吳明")  # print 彭大牆
func("郭靜雅", "王立強", "郭林靜宜", "郭立恆", "林花花")  # print 林花花
func("郭宣雅", "林靜宜", "郭宣恆", "林靜花")  # print 沒有
func("郭宣雅", "夏曼藍波安", "郭宣恆")  # print 夏曼藍波安


# Task 4
def get_number(index):

    if index == 0:
        print(0)
        return 0

    sequence = [0]
    while len(sequence) <= index:
        last_element = sequence[len(sequence) - 1]
        if len(sequence) % 3 == 0:
            sequence.append(last_element - 1)
        else:
            sequence.append(last_element + 4)
    print(sequence[len(sequence) - 1])
    return sequence[len(sequence) - 1]


print("===Task4===")
get_number(1)  # print 4
get_number(5)  # print 15
get_number(10)  # print 25
get_number(30)  # print 70
